### PR TITLE
feat: add auto compaction, fix/refactor conversation state handling

### PR DESCRIPTION
**Pull request by brandonskiser**
*Issue #, if available:* #885

# Overview

This PR intends to accomplish a number of things:
1. Adding auto compaction for the conversation history whenever a context window overflow error is encountered. This is done by summarizing the entire history *excluding the last message* - this is required to ensure that we can make a valid request without causing yet another context window overflow error.
2. Facilitate history handling and support conversation serialization/deserialization by introducing chat specific `UserMessage` and `AssistantMessage` types that wrap `fig_api_client` types. This is required to more appropriately enforce conversation invariants, handle structured context (e.g. to distinguish per-prompt hook context from the user input), and support conversation serialization.
3. Mitigate conversation clearing by increasing the max conversation history length from `100` to `250`, since that is now supported by the API.
4. Correcting the `/compact` calculation
5. Correcting the `/usage` calculation
6. Refactoring much of `ConversationState` to be more type safe

# Details

The original intention of this PR was to add auto compaction, but a sizable refactor was required to accomplish this.

A major problem of `ConversationState` was that we were unable to derive the exact state sent to the model (referenced as `FigConversationState` in the `conversation_state` module). This poses a major problem for `/compact` and `/usage` since we're essentially reimplementing the same calculation over and over for each feature.

This problem is resolved by introducing `BackendConversationState`, an efficient intermediate representation that reflects the exact state sendable to the model at any given point in time.

Other miscellaneous changes to callout:
- Updating `ConversationState::history` to be a list of `(UserInputMessage, AssistantResponseMessage)` instead of `ChatMessage` to add extra type safety around history invariants. (This addresses a bug with the previous compact impl which added multiple assistant messages)
- Adding `TokenCount` and `CharCount` newtypes
- Adding a `SharedWriter` type so that we can have `ChatContext` and `ConversationState` both write to stdout without having to needlessly pass `&mut ChatContext::output` everywhere.
- Changing the `HookExecutor` inline execution logic to directly use `tokio::process` instead due to failing tests in CI (maybe due to `run_command` inheriting stdin?)
- Running hooks on all user messages, rather than just new user requests.

# Followups

1. Gracefully handle the scenario when context files exceed the maximum user message size
2. Add more tests
3. Cache the context file fetching for some short period of time

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: crates/q_cli/src/cli/chat/token_counter.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+#[derive(Debug, Clone, Copy)]
+pub struct CharCount(usize);
+
+impl CharCount {
+    pub fn value(&self) -> usize {
+        self.0
+    }
+}
+
+impl Deref for CharCount {
```
Comment (dingfeli) [inline crates/q_cli/src/cli/chat/token_counter.rs:24]: Why do we need this to be a pointer type.

**Review by dingfeli — APPROVED** [review]
Overall I think it makes sense. 
If there are any problems to be uncovered I think it would have to be done through real world testing. I guess we'll just have to make sure we test this a bunch to cover any edge cases.

File: crates/q_cli/src/cli/chat/conversation_state.rs [hunk 2]
```diff
@@ -173,80 +187,69 @@ impl ConversationState {
...
     pub fn push_assistant_message(&mut self, message: AssistantResponseMessage) {
-        debug_assert!(self.next_message.is_none(), "next_message should not exist");
-        if let Some(next_message) = self.next_message.as_ref() {
-            warn!(?next_message, "next_message should not exist");
+        debug_assert!(self.next_message.is_some(), "next_message should exist");
+        let mut next_user_message = self.next_message.take().expect("next user message should exist");
+
+        // Don't include the tool spec in all user messages in the history.
+        if let Some(ctx) = next_user_message.user_input_message_context.as_mut() {
+            ctx.tools.take();
```
[the remark this answers is not in the document]
Comment (brandonskiser) [reply crates/q_cli/src/cli/chat/conversation_state.rs:197]: We've always removed the tool spec when adding messages to the history, this is just a continuation of that behavior

File: crates/q_cli/src/cli/chat/conversation_state.rs [hunk 3]
```diff
@@ -377,36 +389,165 @@ impl ConversationState {
...
+            .into_fig_conversation_state()
+            .expect("unable to construct conversation state")
+    }
+
+    /// Returns a conversation state representation which reflects the exact conversation to send
+    /// back to the model.
+    pub async fn backend_conversation_state(&mut self, quiet: bool) -> BackendConversationState<'_> {
+        self.enforce_conversation_invariants();
+
+        // Run hooks and add to conversation start and next user message.
```
[the remark this answers is not in the document]
Comment (brandonskiser) [reply crates/q_cli/src/cli/chat/conversation_state.rs:410]: I really want anything that impacts what is stored in the history or sent to the model to derive from a single place, having all of the hook execution logic be passed as an argument to this func just makes testing much more difficult

File: crates/q_cli/src/cli/chat/token_counter.rs (hunk 1 above)
Comment (brandonskiser) [reply crates/q_cli/src/cli/chat/token_counter.rs:24]: It's probably a bad practice - https://stackoverflow.com/questions/45086595/is-it-considered-a-bad-practice-to-implement-deref-for-newtypes

Need to resolve merge conflicts anyways so I can address this as well
